TFUtil.py

```diff
@@ -8517,3 +8517,155 @@ def get_non_deterministic_ops_from_graph():
+                           partition_strategy="mod",
+                           name=None,
+                           seed=None):
+  """Helper function for nce_loss and sampled_softmax_loss functions.
+  Computes sampled output training logits and labels suitable for implementing
+  e.g. noise-contrastive estimation (see nce_loss) or sampled softmax (see
+  sampled_softmax_loss).
+  Note: In the case where num_true > 1, we assign to each target class
+  the target probability 1 / num_true so that the target probabilities
+  sum to 1 per-example.
+
+  This is a copy of
+    https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/ops/nn_impl.py#L1440
+
+  :param tf.Tensor|list(tf.Tensor)|tuple(tf.Tensor) weights: A `Tensor` of shape `[num_classes, dim]`,
+    or a list of `Tensor` objects whose concatenation along dimension 0 has shape `[num_classes, dim]`.
+    The class embeddings.
+  :param tf.Tensor biases: A `Tensor` of shape `[num_classes]`.  The class biases.
+  :param tf.Tensor labels: A `Tensor` of type `int64` and shape `[batch_size, num_true]`.
+    The target classes.  Note that this format differs from
+    the `labels` argument of `tf.nn.softmax_cross_entropy_with_logits`.
+  :param tf.Tensor inputs: A `Tensor` of shape `[batch_size, dim]`.  The forward
+        activations of the input network.
+  :param int num_sampled: The number of classes to randomly sample per batch.
+  :param int num_classes: The number of possible classes.
+  :param int num_true: The number of target classes per training example.
+  :param (tf.Tensor, tf.Tensor, tf.Tensor)|None sampled_values: a tuple of
+    (`sampled_candidates`, `true_expected_count`, `sampled_expected_count`)
+    returned by a `*_candidate_sampler` function.
+    (if None, we default to `log_uniform_candidate_sampler`)
+  :param bool subtract_log_q: whether to subtract the log expected count of
+    the labels in the sample to get the logits of the true labels.
+    Default is True.  Turn off for Negative Sampling.
+  :param bool remove_accidental_hits: Whether to remove "accidental hits"
+    where a sampled class equals one of the target classes.
+  :param str partition_strategy: A string specifying the partitioning strategy, relevant
+    if `len(weights) > 1`. Currently `"div"` and `"mod"` are supported.
+    Default is `"mod"`. See `tf.nn.embedding_lookup` for more details.
+  :param str|None name: A name for the operation.
+  :param int|None seed: random seed for candidate sampling. Default to None, which doesn't set
+    the op-level random seed for candidate sampling.
+  :return:
+    out_logits: `Tensor` object with shape
+        `[batch_size, num_true + num_sampled]`, for passing to either
+        `nn.sigmoid_cross_entropy_with_logits` (NCE) or
+        `nn.softmax_cross_entropy_with_logits` (sampled softmax).
+    out_labels: A Tensor object with the same shape as `out_logits`.
```

Comment:
What is in `out_labels`? Make more clear. Also write the dtype (from "labels", I would infer that it must be integer).